qualtran/bloqs/data_loading/qroam_clean.py

```diff
@@ -0,0 +1,389 @@
+        return tuple(junk_regs)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        ret: Dict[Bloq, SymbolicInt] = defaultdict(lambda: 0)
+        ret[self.qrom_bloq] += 1
+        for swz in self.swap_with_zero_bloqs:
+            if any(is_symbolic(s) or s > 0 for s in swz.selection_bitsizes):
+                ret[swz] += 1
+        return set(ret.items())
+
+    def _build_composite_bloq_with_swz_clean(
+        self,
+        bb: 'BloqBuilder',
+        ctrl: List['SoquetT'],
+        selection: List['SoquetT'],
+        qrom_targets: List['SoquetT'],
+    ) -> Tuple[List['SoquetT'], List['SoquetT'], List['SoquetT']]:
+        sel_l, sel_k = self._partition_sel_register(bb, selection)
+        ctrl, sel_l, qrom_targets = self._add_qrom_bloq(bb, ctrl, sel_l, qrom_targets)
+        sel_k, qrom_targets = self._add_swap_with_zero_bloq(bb, sel_k, qrom_targets)
+        selection = self._unpartition_sel_register(bb, sel_l, sel_k)
+        return ctrl, selection, qrom_targets
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        # Get the ctrl and target register for the SelectSwapQROM.
+        ctrl = [soqs.pop(reg.name) for reg in self.control_registers]
+        selection = [soqs.pop(reg.name) for reg in self.selection_registers]
+        if is_symbolic(*self.block_sizes):
+            raise ValueError(
+                f"Cannot decompose QROAM bloq with symbolic block sizes. Found {self.block_sizes=}"
+            )
+        block_sizes = cast(Tuple[int, ...], self.block_sizes)
+        # Allocate intermediate clean/dirty ancilla for the underlying QROM call.
+        qrom_targets = []
+        for reg in self.target_registers:
+            qrom_target = _alloc_anc_for_reg_except_first(
+                bb, reg.dtype, block_sizes, self.use_dirty_ancilla
+            )
+            qrom_target[np.unravel_index(0, block_sizes)] = _alloc_anc_for_reg(  # type: ignore[index]
+                bb, reg.dtype, reg.shape, dirty=False
+            )
+            qrom_targets.append(qrom_target)
+        # Assert that all registers have been used by now.
+        assert not soqs, f"All registers must have been used by now. Found: {soqs}"
+        # Add the bloq decomposition
+        if any(b > 1 for b in block_sizes):
+            ctrl, selection, qrom_targets = self._build_composite_bloq_with_swz_clean(
+                bb, ctrl, selection, qrom_targets
+            )
+        else:
+            ctrl, selection, qrom_targets = self._add_qrom_bloq(bb, ctrl, selection, qrom_targets)
+        # Construct and return dictionary of final soquets.
+        soqs |= {reg.name: soq for reg, soq in zip(self.control_registers, ctrl)}
+        soqs |= {reg.name: soq for reg, soq in zip(self.selection_registers, selection)}
+        soqs |= {reg.name: soq.flat[1:] for reg, soq in zip(self.junk_registers, qrom_targets)}  # type: ignore[union-attr]
```

Comment:
Could you describe this convention somewhere? Maybe above where qrom_targets are built.

Reply:
Yes, the idea is that all the `k-1` newly allocated registers are right registers and are kept around as junk. The adjoint bloq expects has the same API and will be responsible to deallocate the target + junk registers using measurement based uncomputation. I've also added some figures as part of the PR description to make it easier to see what's going on. 

The classical simulation is also implemented in a way that for each selection index `s`, the correct permutation of `target + junk` registers is computed and returned; and the adjoint bloq is instantiated so that uncomputing the original qroam read reduces to a  table lookup of size `N / 2` and target bitsize `b * k` -- the important thing to note is that cost of uncomputation does not dependent upon the target bitsize (`b * k`) and this is reflected in the costs; but the classical simulation tests assert that adjoint bloq has the correct table as input.